<h1>Objektově orientované programování</h1>
<p>Objektově orientované programování je přístup k řešení problému pomocí takzvaných objektů.</p>


<h2>Syntaxe a sémantika</h2>
<p><pre>class nazev_tridy:
  def nazev_metody(self, arg1, arg2):
    prikaz1
    prikaz2
    prikaz3

  def nazev_dalsi_metody(self):
    prikaz4

O1 = nazev_tridy()
</pre></p>
<p>Třídu vytvoříme pomocí klíčového slova <code>class</code>. Tím jsme ještě <strong>NEvytvořili</strong> žádný objekt.<br>Uvnitř třídy pak píšeme jednotlivé metody (dříve jsme jim říkali funkce), které potom můžeme u objektů této třídy využít.<br>Objekt vytvoříme zavoláním jména třídy, jako bychom volali funkci (tedy s kulatými zavorkami)</p>

<h2>Objekt, atributy</h2>
<p>Objekt je samostatná "jednotka" -> tedy pokud máme takových objektů více (např. více hráčů ve hře), je každý z nich nezávislý na ostatních (každý je na jiném místě, má jiné skóre, jiné HP, ...). Těmto vlastnostem (pozice, hp, skóre) říkame <strong>atributy</strong></p>
<h2>class - Třída</h2>
<p>Třídu si můžeme představit jako „šablonu“ nebo „výrobní plán“ , podle kterého se vytváří jednotlivé objekty - říkame jim objekty dané třídy. V ku ji vytvoříme pomocí klíčového slova <code>class</code>.<br>Ve chvíli, kdy v programu zavoláme příkaz pro vytvoření objektu třídy, vytvoří se objekt, který má atributy (např. pozice, skóre, HP) a metody (pohyb, útok, ...), které jsme mu předepsali ve třídě.</p>

<h2>Metody</h2>
<p>Metody (dříve jsme jim říkali "funkce" - je to jen jiné slovo s velmi podobným významem) jsou "dovednosti", které můžeme u objektu využívat.<br>Tyto "dovednosti" (metody) musíme objekty dané třídy nejprve "naučit" (napsat metodu, které provede požadované operace).<br>Výhodné je, že metodu zapíšeme do třídy (tedy pouze jednou pro všechny objekty dané třídy) a poté ji můžeme využít u všech objektů.</p>

<h2>_ _init_ _()</h2>
<p>Metoda <code>__init__()</code> je speciální metoda (speciální metody jsou označny dvěma podtržítky), která se zavolá ve chvíli, kdy vytváříme objekt.<br>
Při vytváření objektu se vykonají všechny příkazy v této metodě a stejně jako každá jiná metoda umí pracovat i s argumenty. Argumenty využijeme pro nastavení počátečních hodnot objeku.</p>


<h2>self</h2>
<p>V OOP kódu velmi často uvidíte slovo <code>self</code>.<br>Je to proměnná, ve které je uložená adresa objektu se kterým aktuálně pracujeme.<br>
Pokud máme v programu tři objekty Auto - Škodovka, Trabant, Ferrari a zavoláme například <code>jed_doprava()</code> na objekt Trabant. Dává smysl, že se pohne pouze tento jeden objekt -> který ze tří objektů to je (na který objekt jsme metodu zavolali), musí mít Python někde uložené - a má to uložené právě v proměnné <code>self</code>.</p>

In [ ]:
# vytvořením třídy jsme NEvytvořili žádný objekt
class Auto: # třída ("šablona") pro objekty Auto
  def __init__ (self, x, y, nazev): # tato metoda se spustí při vytváření objektu
    self.pozice = [x,y]
    self.hp = 0
    self.nazev = nazev

  def vypisInformace(self):
    return "Auto {} s hp {} na pozici {}".format(self.nazev, self.hp, self.pozice)
  def __str__(self): # speciální metoda __str__() určuje, jak se objekt převede na string
    return "Auto {} s hp {} na pozici {}".format(self.nazev, self.hp, self.pozice)

auto1 = Auto(0,0,"Škodovka") # zde vytváříme objekt typu Auto - tedy spouštíme metodu __init__() ve třídě Auto
auto2 = Auto(10,20,"Trabant")
auto3 = Auto(300,400,"Ferrari")

print(auto1, type(auto1)) # funkce print() vždy tiskne string - zavolá se tedy metoda __str__() a vrácený string se zobrazí
print(auto2, type(auto2))
print(auto3, type(auto3))        aaaa

Auto Škodovka s hp 0 na pozici [0, 0] <class '__main__.Auto'>
Auto Trabant s hp 0 na pozici [10, 20] <class '__main__.Auto'>
Auto Ferrari s hp 0 na pozici [300, 400] <class '__main__.Auto'>


<h2>Přečtení hodnot</h2>
<p>Objektu se samozřejmě často potřebujeme zeptat, jaké jsou hodnoty jeho atributů (jaký má název, kolik má hp, jaká je jeho pozice).<br>K hodnotám atributů se dostaneme pomocí <code>.</code> (tečky).</p>

In [ ]:
# pro spuštění této buňky musíte předtím alespoň jednou spustit buňku, ve které se vytváří objekty auto1, auto2 a auto3
print("Auto1: x:", auto1.pozice[0],"y:", auto1.pozice[0], "hp:", auto1.hp, "nazev:", auto1.nazev)
print("Auto2: x:", auto2.pozice[0],"y:", auto2.pozice[1], "hp:", auto2.hp, "nazev:", auto2.nazev)
print("Auto3: x:", auto3.pozice[0],"y:", auto3.pozice[1], "hp:", auto3.hp, "nazev:", auto3.nazev)

Auto1: x: 0 y: 0 hp: 0 nazev: Škodovka
Auto2: x: 10 y: 20 hp: 0 nazev: Trabant
Auto3: x: 300 y: 400 hp: 0 nazev: Ferrari


<h1>Cvičení</h1>

<h2>1</h2>
<p>Vytvořte třídu <code>Zeton</code>.<br>Při vytváření objektu této třídy zadáme tři hodnoty - <code>pozice_x</code>, <code>pozice_y</code> a <code>hodnota</code> (kolik bodů hráči přinese sebrání tohoto žetonku).<br>Tyto tři hodnoty si bude <code>Zeton</code> také pamatovat - v atributu <code>pozice</code> typu List bude <code>[pozice_x, pozice_y]</code> a v atributu <code>hodnota</code> typu int bude uložena jeho hodnota.</p>

In [ ]:
class Zeton:
    def __init__(self,pozice_x: int,pozice_y: int,hodnota):
        self.pozice = [pozice_x,pozice_y]
        self.hodnota = hodnota
z1 = Zeton(1,-2,8)

vsechny_zetony = []
for i in range(0,5):
    vsechny_zetony.append(Zeton(-i 10*i, 7*i))

print("existuje {} žetonu:".format(len(vsechny_zetony)))
for i, z in enumerate(vsechny_zetony):
    print("{}. pozice: {} hodnota {}".format(i+1, z.pozice, z.hodnota))

Existuje 5 Žetonů:
1. pozice: [0, 0] hodnota 0
2. pozice: [-1, 10] hodnota 7
3. pozice: [-2, 20] hodnota 14
4. pozice: [-3, 30] hodnota 21
5. pozice: [-4, 40] hodnota 28


<p>Výsledky pro Vaší kontrolu - referenční řešení má 14 řádků</p>
<table>
<tr>
<td>
<pre>
Existuje 5 Žetonů:
1. pozice: [0, 0] hodnota 0
2. pozice: [-1, 10] hodnota 7
3. pozice: [-2, 20] hodnota 14
4. pozice: [-3, 30] hodnota 21
5. pozice: [-4, 40] hodnota 28
</pre>
</tr>
</table>

<h2>2</h2>
<p>Upravte svůj předchozí kód tak, aby pozice Žetonu nemohla být záporná (tedy byla 0 a více). Pokud se pokusíme vytvořit Žeton se zápornou pozicí, bude jeho pozice 0</p>

In [ ]:
# TODO: zde vytvořte třídu
class Zeton:
  def __init__(self,pozice_x,pozice_y,hodnota):
    if pozice_x < 0:
      pozice_x = 0
    if pozice_y < 0:
      pozice_y = 0
    self.pozice = [pozice_x,pozice_y] #také lze self.pozice = [max(0,pozice_x),max(0,pozice_y)]
    self.hodnota = hodnota

# Dále už kód neměňte
vsechny_zetony = []
for i in range(0,5):
  vsechny_zetony.append(Zeton(2-i, 10*i, 7*i))

print("Existuje {} Žetonů:".format(len(vsechny_zetony)))
for i, z in enumerate(vsechny_zetony):
  print("{}. pozice: {} hodnota {}".format(i+1, z.pozice, z.hodnota))

Existuje 5 Žetonů:
1. pozice: [2, 0] hodnota 0
2. pozice: [1, 10] hodnota 7
3. pozice: [0, 20] hodnota 14
4. pozice: [0, 30] hodnota 21
5. pozice: [0, 40] hodnota 28


<p>Výsledky pro Vaší kontrolu - referenční řešení má 14 řádků</p>
<table>
<tr>
<td>
<pre>
Existuje 5 Žetonů:
1. pozice: [2, 0] hodnota 0
2. pozice: [1, 10] hodnota 7
3. pozice: [0, 20] hodnota 14
4. pozice: [0, 30] hodnota 21
5. pozice: [0, 40] hodnota 28
</pre>
</td>
</tr>
</table>

<h2>3</h2>
<p>Do třídy <code>Auto</code> přidejte metodu <code>jed_vlevo(self)</code>.<br>
Pokud zavoláme tuto metodu na objekt, zmenší se jeho pozice na x-ové ose o 5.</p>


In [ ]:
class Auto: # třída ("šablona") pro objekty Auto
  def __init__ (self, x, y, nazev): # tato metoda se spustí při vytváření objektu
    self.pozice = [x,y]
    self.hp = 0
    self.nazev = nazev

  def __str__(self): # speciální metoda __str__() určuje, jak se objekt převede na string
    return "Auto {} s hp {} na pozici {}".format(self.nazev, self.hp, self.pozice)

  # TODO: zde vytvořte metodu
  def jed_vlevo(self):
    self.pozice[0] = self.pozice[0]-5 #   self.pozice[0] -= 5

# Dále už kód neměňte
auto3 = Auto(300,400,"Ferrari")
for i in range(5):
  print(auto3)
  auto3.jed_vlevo()

print("Auto na konci:", auto3)

Auto Ferrari s hp 0 na pozici [300, 400]
Auto Ferrari s hp 0 na pozici [295, 400]
Auto Ferrari s hp 0 na pozici [290, 400]
Auto Ferrari s hp 0 na pozici [285, 400]
Auto Ferrari s hp 0 na pozici [280, 400]
Auto na konci: Auto Ferrari s hp 0 na pozici [275, 400]


<p>Výsledky pro Vaší kontrolu - referenční řešení má 20 řádků</p>
<table>
<tr>
<td>
<pre>
Auto Ferrari s hp 0 na pozici [300, 400]
Auto Ferrari s hp 0 na pozici [295, 400]
Auto Ferrari s hp 0 na pozici [290, 400]
Auto Ferrari s hp 0 na pozici [285, 400]
Auto Ferrari s hp 0 na pozici [280, 400]
Auto na konci: Auto Ferrari s hp 0 na pozici [275, 400]
</pre>
</td>

</tr>
</table>

<h2>4</h2>
<p>Upravte vaší metodu <code>jed_vlevo(self)</code> z předchozího cvičení tak, aby objekt nemohl "vjet" do záporných čísel - pokud by se to mělo stát, "zasekne" se na 0.</p>

In [ ]:
class Auto: # třída ("šablona") pro objekty Auto
  def __init__ (self, x, y, nazev): # tato metoda se spustí při vytváření objektu
    self.pozice = [x,y]
    self.hp = 0
    self.nazev = nazev

  def __str__(self): # speciální metoda __str__() určuje, jak se objekt převede na string
    return "Auto {} s hp {} na pozici {}".format(self.nazev, self.hp, self.pozice)

  # TODO: zde vytvořte metodu
  def jed_vlevo(self):
    # self.pozice[0]= max(0,self.pozice[0]-5)
    if self.pozice[0]-5 < 0:
      self.pozice[0] = 0
    else:
      self.pozice[0] = self.pozice[0]-5

# Dále už kód neměňte
auto3 = Auto(10,400,"Ferrari")
for i in range(5):
  print(auto3)
  auto3.jed_vlevo()

print("Auto na konci:", auto3)

Auto Ferrari s hp 0 na pozici [10, 400]
Auto Ferrari s hp 0 na pozici [5, 400]
Auto Ferrari s hp 0 na pozici [0, 400]
Auto Ferrari s hp 0 na pozici [0, 400]
Auto Ferrari s hp 0 na pozici [0, 400]
Auto na konci: Auto Ferrari s hp 0 na pozici [0, 400]


<p>Výsledky pro Vaší kontrolu - referenční řešení má 20 řádků</p>
<table>
<tr>
<td>
<pre>
Auto Ferrari s hp 0 na pozici [10, 400]
Auto Ferrari s hp 0 na pozici [5, 400]
Auto Ferrari s hp 0 na pozici [0, 400]
Auto Ferrari s hp 0 na pozici [0, 400]
Auto Ferrari s hp 0 na pozici [0, 400]
Auto na konci: Auto Ferrari s hp 0 na pozici [0, 400]
</pre>
</td>

</tr>
</table>

# 5

Založte třídu Byt, která bude mít atributy:

* velikost (4kk)
* x (rozmer jednoho směru/ stěny)
* y (rozmer druhého směru/ stěny)
* vlastnik

A bude obsahovat metody init, toString a metodu rozmeryVmetrech, která vám vrátí rozměry bytu v m^2

Vytvořte byt a poté si nechte vypsat:
* Vše o bytě
* jaká je rozloha bytu
* Kdo je vlastník

In [ ]:
class Byt:
    def __init__(self,velikost,x,y,vlastnik):
        self.velikost = velikost
        self.x = x
        self.y = y
        self.vlastnik = vlastnik

    def rozmeryVmetrech(self):
        return self.x * self.y

    def rozmeryVmetrechToString(self):
        print("Rozloha bytu je:",self.x * self.y, "m^2")

    def __str__(self):
        return f"Byt o dispozici {self.velikost} a rozměrech {self.x}x{self.y} což je: {self.rozmeryVmetrech()} vlastní {self.vlastnik}"

byt = Byt("4kk",25,30,"Karel")
byt2 = Byt("2kk",20,80,"Jarda")

print(byt)
print("Rozloha bytu je:", byt.rozmeryVmetrech(), "m^2")
print("Vlastníkem je: ", byt.vlastnik)
byt.rozmeryVmetrechToString()

Byt o dispozici 4kk a rozměrech 25x30 vlastní Karel
Rozloha bytu je: 750 m^2
Vlastníkem je:  Karel
